base_layer/core/src/chain_storage/blockchain_database.rs

```diff
@@ -1756,105 +1732,109 @@ fn insert_orphan_and_find_new_tips<T: BlockchainBackend>(
-        new_tips_found.push(new_tip);
+    let tips = find_orphan_descendant_tips_of(&*db, &chain_header, validator, &mut txn)?;
+    debug!(target: LOG_TARGET, "Found {} new orphan tips", tips.len());
+    for new_tip in &tips {
+        txn.insert_orphan_chain_tip(new_tip.hash().clone());
     }
 
     db.write(txn)?;
-    Ok(new_tips_found)
+    Ok(tips)
 }
 
 // Find the tip set of any orphans that have hash as an ancestor
 fn find_orphan_descendant_tips_of<T: BlockchainBackend>(
     db: &T,
-    prev_header: &BlockHeader,
-    prev_accumulated_data: &BlockHeaderAccumulatedData,
+    prev_chain_header: &ChainHeader,
     validator: &dyn HeaderValidation<T>,
     txn: &mut DbTransaction,
 ) -> Result<Vec<ChainHeader>, ChainStorageError>
 {
-    let children = db.fetch_orphan_children_of(prev_accumulated_data.hash.clone())?;
+    let children = db.fetch_orphan_children_of(prev_chain_header.hash().clone())?;
     if children.is_empty() {
-        return Ok(vec![ChainHeader {
-            header: prev_header.clone(),
-            accumulated_data: prev_accumulated_data.clone(),
-        }]);
+        debug!(
+            target: LOG_TARGET,
+            "Found new orphan tip {} ({})",
+            prev_chain_header.height(),
+            prev_chain_header.hash().to_hex()
+        );
+        return Ok(vec![prev_chain_header.clone()]);
     }
+
     let mut res = vec![];
     for child in children {
-        match validator.validate(db, &child.header, prev_accumulated_data) {
-            Ok(builder) => {
-                let accum_data = builder
-                    .total_kernel_offset(
-                        &prev_accumulated_data.total_kernel_offset,
-                        &child.header.total_kernel_offset,
-                    )
+        match validator.validate(db, &child.header) {
+            Ok(achieved_target) => {
+                let child_hash = child.hash();
+                let accum_data = BlockHeaderAccumulatedData::builder(prev_chain_header.accumulated_data())
+                    .with_hash(child_hash.clone())
+                    .with_achieved_target_difficulty(achieved_target)
+                    .with_total_kernel_offset(child.header.total_kernel_offset.clone())
                     .build()?;
-                txn.insert_chained_orphan(
-                    ChainBlock {
-                        block: child.clone(),
-                        accumulated_data: accum_data.clone(),
-                    }
-                    .into(),
-                );
-                res.extend(find_orphan_descendant_tips_of(
-                    db,
-                    &child.header,
-                    &accum_data,
-                    validator,
-                    txn,
-                )?);
+
+                let chain_header = ChainHeader::try_construct(child.header, accum_data)
+                    .expect("program correctness: accumulated data did not match the given block hash");
```

Comment:
I think this should be propagated. There is still a small chance this will happen and it will be easier to find in the logs than a panic

Reply:
As it stands now, it is impossible because the child hash is set to the accumulated data hash (i.e. you would have to make a programming error to trigger this panic). That being said, if that happens the logs should make it pretty clear. Will change it to an error .